### ISSUE TITLE
[FEATURE_REQUEST] Trimming the chat history in chunks to speed up prompt ingestion for llama.cpp

### ISSUE THREAD

**Issue by h3ndrik** [SillyTavern/SillyTavern#1278]
### Have you searched for similar requests?

Yes

### Is your feature request related to a problem? If so, please describe.

llama.cpp has the feature to re-use the context window if the beginning of it doesn't change.

I've always had the problem that llama.cpp needs to re-compute the whole context window after it gets filled and old chat messages are trimmed one by one. At that point most of the prompt changes on every chat message, forcing it to re-ingenst thousands of tokens on every new chat message.

### Describe the solution you'd like

I propose we allow users to trim out larger chunks out of the chat history (instead of a bit every iteration) so these re-computations happen more infrequently. An option to give a lower bound that the chat history gets trimmed to (once it exceeds the upper bound) would do it. Or an option that specifies the size of the chunk that gets deleted.

### Additional context

As a further feature that could be tied into the summarization. Triggering a summary of the deleted text that would be also inserted at that point where the whole context needs to be re-computed anyways.

### Priority

Low (Nice-to-have)

### Is this something you would be keen to implement?

No

**Comment by HiroseKoichi** [comment]
The prompt is already automatically truncated to fit the context length, which is why you don't immediately run out of memory with a long chat history.
The issue you're referring to has only recently been solved: https://github.com/mit-han-lab/streaming-llm
And a relevant thread on llama.cpp can be found here: https://github.com/ggerganov/llama.cpp/issues/3440

**Comment by h3ndrik** [comment]
> The issue you're referring to has only recently been solved [...]

Yeah, that won't work. The character card needs to stay at the front or the AI will forget the character traits. Then there is an incision and the chat history after that is a rolling window. But we'd need everything rolling through the context for that to work.

That works when generating text or stories. But you need to be willing to delete/phase out the beginning (the loaded character card) implicitly once the context is filled. So it's impossible for something like SillyTavern to use the approach from the paper. 

> The prompt is already automatically truncated to fit the context length [...]

I want it truncated in a different way (in chunks). The way it currently works is exactly why it's unnecessarily slow on llama.cpp.

**Comment by HiroseKoichi** [comment]
> Yeah, that won't work. The character card needs to stay at the front or the AI will forget the character traits. Then there is an incision and the chat history after that is a rolling window. But we'd need everything rolling through the context for that to work.
> 
> That works when generating text or stories. But you need to be willing to delete/phase out the beginning (the loaded character card) implicitly once the context is filled. So it's impossible for something like SillyTavern to use the approach from the paper.

You seem to be misunderstanding how streaming-LLM works. The example that they've shown on their github is to demonstrate that re-computation doesn't need to happen when readjusting what's in the context window. This doesn't in any way affect the character card format, nor does it require it to be one continuous prompt.

Here is a link to a video explaining what this is and what it does: https://www.youtube.com/watch?v=409tNlaByds

> I want it truncated in a different way (in chunks). The way it currently works is exactly why it's unnecessarily slow on llama.cpp.

There are no different ways of truncating the prompt; you just trim it to fit the context window. If you have 8000 tokens of chat history and your context window is 4096, then the prompt is trimmed to be 4096 tokens before being fed to the LLM. We're not feeding it the full 8000 tokens and then having it compute all of them only to use 4096 tokens.

When the prompt exceeds the context window, it will always need to be recomputed. The reason you don't see these slowdowns when the context window isn't full is because it's using KV cache to cache already existing prompts, but as soon as it has to drop out the old context and thus KV cache, it has to re-compute the prompt with the new information included.

**Comment by h3ndrik** [comment]
Yeah, but we're not readjusting the context window here. the StreamingLLM works for cases like this (numbers being the tokens / chat history that gets swapped out of the window):
`[ 1 2 3 4 5 6 7 ]` -> `[ 2 3 4 5 6 7 8 ]` -> `[ 3 4 5 6 7 8 9 ]`

we have this:

`[ c c c 1 2 3 4 ]` -> `[ c c c 2 3 4 5 ]` -> `[ c c c 3 4 5 6 ]`

With the 'c' being the character description and the story and surroundings. That is fixed. And I don't see how that wouldn't be a different prompt every time (to StreamingLLM).

> [...] but as soon as it has to drop out the old context and thus KV cache, it has to re-compute the prompt with the new information included.

And it happens practically on every new message once it's full. My proposal is: Truncate out like 1024 tokens at once. And now your context is only 3072 tokens and you have quite some time in which you can re-use the KV-cache until it hits 4096 again and the next truncation needs to happen:

`[ c c 1 2 3 4 _ ]` (cache)-> `[ c c 1 2 3 4 5 ]` (re-compute)-> `[ c c 4 5 6 _ _ ]` (cache)-> `[ c c 4 5 6 7 _ ]` (cache)-> `[ c c 4 5 6 7 8 ]`

vs

`[ c c 1 2 3 4 _ ]` (cache)-> `[ c c 1 2 3 4 5 ]` (re-compute)-> `[ c c 2 3 4 5 6 ]` (re-compute)-> `[ c c 3 4 5 6 7 ]` (re-compute)-> `[ c c 4 5 6 7 8 ]`

**Comment by HiroseKoichi** [comment]
Once again, you seem to be misunderstanding how streaming-LLM works. The reordering of the tokens is just one part of the proposal, not the main focus. And it doesn't negatively affect the character card descriptions.

There are three main aspects of streaming-LLM: attention sink, sliding window attention, and token reordering.

Attention sink: What the authors of streaming-LLM have found is that most of the attention is focused on the first few tokens because it's dumping the excess attention onto them. So the idea is to keep the first few tokens, or in their proposal, retrain models with a null token where all the excess attention will be focused. This means that the most computationally heavy part can essentially be copied and pasted.

Sliding window attention: The current behavior is that all tokens attend to themselves and every other token in the prompt. Sliding window attention makes it so the attention only goes one way, with the most recent tokens attending to the most and the ones at the beginning only attending to a few. This is very important to streaming-LLM because now the KV cache can be used continuously instead of recomputing a new KV cache for the whole prompt every time new information is added (when combined with the two other things), because the old information is no longer attending to future information, so we can safely use the existing KV cache.

Token reordering: The current behavior is to shift the token position when outside of the context window. For example, if you have a context window of 10 and a prompt length of 15, then the LLM will see it as 6,7,8,9,10,11,12,13,14,15. And what they are proposing is to tell the LLM that the 6th token is actually the 1st, and the 15th is actually the 10th. This is so it can smoothly and effectively use the KV cache. It doesn't drop out or change any information in the prompt.

I highly recommend you watch this video to actually understand the topic of what we're discussing instead of just making assumptions about what it does: https://www.youtube.com/watch?v=409tNlaByds

> And it happens practically on every new message once it's full. My proposal is: Truncate out like 1024 tokens at once. And now your context is only 3072 tokens and you have quite some time in which you can re-use the KV-cache until it hits 4096 again and the next truncation needs to happen:

What you're proposing here doesn't solve anything; you still have to recompute every time information is dropped out; you're just reducing the context window for little to no gain.

> The example that they've shown on their github is to demonstrate that re-computation doesn't need to happen when readjusting what's in the context window.

> Yeah, but we're not readjusting the context window here.

I never said we were adjusting the context window itself; I said 'readjusting what's in the context window'.

**Comment by h3ndrik** [comment]
> you seem to be misunderstanding how streaming-LLM works

Sorry. I kinda get how it works but seems I was too focused on how it works as of now in SillyTavern. I can re-watch that video once I got some time to spare. But my main point for this feature-request is how it applies to SillyTavern.

Sure we could just append the dialogue to a infinite context, use [Streaming-LLM](https://github.com/mit-han-lab/streaming-llm) and never truncate the chat history. Let the LLM choose what it attends to. I somehow completely missed that.

But a infinite context size isn't implemented in SillyTavern either.

And not (yet) implemented in llama.cpp and not on the roadmap ([sliding window attention](https://github.com/ggerganov/llama.cpp/issues/3377) is on the roadmap, though).

And it seems Streaming-LLM is not available in other major inference frameworks either.

And we can't change the prompt in the middle, only at the end. So extensions like 'summary' or a vector database won't work along Streaming-LLM. (You'd need to assure that or there's at one point like 30.000 tokens to re-ingest.)

> What you're proposing here doesn't solve anything; you still have to recompute every time information is dropped out; you're just reducing the context window for little to no gain.

Yes it does. I've actually implemented in one of my early tries with Langchain. Back in the day we only had 2048 tokens of context and it wasn't feasible to drop large chunks. But now that we universally have 4096+ tokens, we can easily trim out (yes, this lowers it's effective size) a thousand tokens. It does away with like 9/10 of the heavy re-computations until these thousand tokens are filled again. Yielding a 90% performance increase.

For me llama.cpp keeps the static character description in the kv-cache most times. It re-computes the ~3000 tokens that get trimmed and follow the character description. On my machine that takes well more than a minute. If I write like 20 messages onto a filled chat history, thats like 20\*2 minutes = 40 minutes of re-computation and wait time.
With my approach I lower the context size (I still have like 2000 tokens minimum for the chat history) but I now have 1 recomputation, 9 messages that happen fast (in 15s) then the context is filled and i need to recompute once (2m) and then have 9 messages fast again. That's 2\*2minutes + 18\*15seconds = 8:30 minutes.

So with these (arbitrary) numbers it increases the total wait time for 20 messages in a filled context from 40 minutes to 8,5 minutes. I'd say this is a huge increase. And it's relatively easy to implement.

I don't say this is the only solution. But a good compromise that's available right now with little effort. Until things like Streaming-LLM are universally available.

(And I really have to have another look at Streaming-LLM. It's absolutely necessary for SillyTavern that it attends to the token in the way they found out. We need attention to the beginning of the context where the character description is located and to the end where the most relevant lines of dialogue are. The total amount of attention is fixed, so it really needs to attend in the way that aligns well to the way the "Story String" formats the prompt. But seems that is well possible. But I need to look at the benchmarks in the paper.)

**Comment by Cohee1207** [comment]
> My proposal is: Truncate out like 1024 tokens at once.

I took some time to think about this, but I have no idea how to translate your request into the actual code. The message buffer is filled in reverse order until it stops fitting the context, and the length of every message is unpredictable. Even in the hypothetical case when the sacrifice has been made and the prompt builder is fully rewritten to accommodate for the llama.cpp quirks, the cache invalidation won't go away completely, so switching chats will still cost you time and it will stay a problem until someone finally discovers how to get rid of it on the fundamental level.

**Comment by h3ndrik** [comment]
> the cache invalidation won't go away completely

Thanks for commenting. That is true. I can't do away with the re-computations altogether, I can only make them less frequent. It needs to happen every so often (without infinite context). However, for my main way of using SillyTavern, it'd be a huge speed-up.

> switching chats will still cost you time and it will stay a problem until someone finally discovers how to get rid of it on the fundamental level

True. I don't see a 'fundamental level' solution. llama.cpp could cache multiple chats in parallel, that would solve this at the expense of quite some RAM. But the backend (llama.cpp) is probably the wrong place to do pattern matching and try to guess how to do caching for diverse frontends. ([KoboldCPP](https://github.com/LostRuins/koboldcpp) does something similar with '--smartcontext'. But that triggers rarely and is only one cache.)

Switching chats and things alike trigger a whole re-computation and will not benefit from this. I'd argue most of the user's total actions are normal dialogue, replying and not sending 1 message to each chatbot in succession. So it would be beneficial to the main thing a user does. But not to everything, granted.

> accommodate for the llama.cpp quirks

I don't really know how many people use llama.cpp on a CPU. And if this is a niche use-case or something that could make a few more people happy. AFAIK on GPU people don't care about efficiency too much. The KV-cache is a unique feature of llama.cpp and most of the people just re-compute the whole prompt each time because it's fast enough.

> The message buffer is filled in reverse order until it stops fitting the context

Yeah, I had a quick look at the code and that's why I opened a feature-request instead of patching it myself. I'm not sure if people would like to change that. We either need to count tokens and change the way the buffer gets filled, or do some sort of a cheap compromise like trimming 20% of the buffer by number of dialogue lines or something like that. (But that'd be sub-par and probably require more fine-tuning by the user.)

**Comment by HiroseKoichi** [comment]
> Sure we could just append the dialogue to a infinite context, use [Streaming-LLM](https://github.com/mit-han-lab/streaming-llm) and never truncate the chat history. Let the LLM choose what it attends to. I somehow completely missed that.
> 
> But a infinite context size isn't implemented in SillyTavern either.

Streaming-LLM doesn't give an infinite context size; what it does is get rid of the prompt processing delay when the context fills up. The video on their github is to demonstrate that even when the context fills up and the attention slides over, there are no performance drops and it can keep generating like normal, unlike the video on the left, which shows a pause in generation as the context fills up followed by an out-of-memory error. Having it continue generating forever was just an extreme example to demonstrate its effectiveness.

Streaming-LLM doesn't change how we would feed the LLM information or how we use it; it only gets rid of the prompt processing delay, allowing you to generate instantly instead of waiting for it to process the prompt. Though there is a positive side effect of streaming-LLM, since we aren't tossing out the old KV cache and generating new ones, we can get a sort of daisy chain effect where, while new tokens can't attend to tokens outside of the context window, we can still attend to tokens that have. This means we might potentially get a pseudo context size increase, but we'll have to wait and see once we can run it with all the regular tools.

> And not (yet) implemented in llama.cpp and not on the roadmap ([sliding window attention](https://github.com/ggerganov/llama.cpp/issues/3377) is on the roadmap, though).
> 
> And it seems Streaming-LLM is not available in other major inference frameworks either.

I agree. It will likely take a while for even just sliding window attention to be implemented, but that's just the way things are; there's a lot of rapid progress, but it takes time to implement these things, especially when FOSS projects are usually led by small teams that aren't being paid to do it (most of the time).

> I don't really know how many people use llama.cpp on a CPU. And if this is a niche use-case or something that could make a few more people happy. AFAIK on GPU people don't care about efficiency too much. The KV-cache is a unique feature of llama.cpp and most of the people just re-compute the whole prompt each time because it's fast enough.

I haven't seen very many people say they run CPU only, but that could be because people don't really give details about their setup in normal conversation, so it could be much higher than I think. A lot of people that I've seen have been using paid APIs like OpenAI or Claude, and the ones running them locally tend to stick with 7B models so they can run them with good performance.

I don't know what parameters of models you run, but I recommend Mistral-7B fine-tunes. Even though I have a 6700xt, they're my main models at the moment; they outperform 13Bs, the 30Bs are based on code llama and give even worse results than the 13Bs, and the 70Bs are just a bit too slow for my liking. In addition, Mistral has a native 8K context that can go up to 32K with sliding window attention.

My favorite Mistral fine-tune is https://huggingface.co/TheBloke/dolphin-2.1-mistral-7B-GGUF at the moment, but lots more models have been released, so I still have a lot to test. You should be able to run Q8 with decent performance since Mistral uses GQA, and at the very least it should be faster than what you were using. But if you do decide to use it, just keep in mind that the ChatML presets don't work properly at the moment, so something like https://huggingface.co/TheBloke/SynthIA-7B-v2.0-GGUF would probably perform better. I haven't tested the new Synthia 2.0 or even 1.5 yet, but they have been really solid so far.

**Comment by h3ndrik** [comment]
> Streaming-LLM doesn't give an infinite context size;

From their paper:

"Based on the above insights, we propose StreamingLLM, a simple and efficient framework that
enables LLMs trained with a finite attention window to **work on text of infinite length** without finetuning."

---

Let's not discuss Streaming-LLM here. Let's keep at the issue at hand. Thanks for the model suggestions. I've had some success with [Mistral 7B Claude](https://huggingface.co/Norquinal/Mistral-7B-claude-chat) and [Mistral 7B OpenOrca](https://huggingface.co/Open-Orca/Mistral-7B-OpenOrca). But I'm not 100% up-to-date. I'm going to try the ones you mentioned. I still like some of the Llama2 13B fine-tunes, depending on the use-case. I use Q4_K_M quantization (sometimes Q5_K_M). I think it's as close to the sweet-spot as we can currently get.

**Comment by HiroseKoichi** [comment]
> > Streaming-LLM doesn't give an infinite context size;
> 
> From their paper:
> 
> "Based on the above insights, we propose StreamingLLM, a simple and efficient framework that enables LLMs trained with a finite attention window to **work on text of infinite length** without finetuning."

You've misunderstood the sentence; context size =/= text length. Nowhere do they mention infinite context size here, though that would be a massive breakthrough.

A model will have its output degraded when you try to run it on a text length that exceeds its trained context size. For example, if you extend a 4k context model to 8k, then it will have its output degraded. What they're saying is that streaming-LLM allows a model to work with text lengths that exceed its trained context length with no output degradation.

On literally the next page of the paper:

**Extensive research has been done on applying LLMs to lengthy texts, with three main areas of focus:
Length Extrapolation, Context Window Extension, and Improving LLMs’ Utilization of Long
Text.** While seemingly related, it’s worth noting that progress in one direction doesn’t necessarily
lead to progress in the other. For example, extending the context size of LLMs doesn’t improve the
model’s performance beyond the context size, and neither approach ensures effective use of the long
context. **Our StreamingLLM framework primarily lies in the first category, where LLMs are applied
to text significantly exceeding the pre-training window size, potentially even of infinite length. We do
not expand the attention window size of LLMs or enhance the model’s memory and usage on long
texts. The last two categories are orthogonal to our focus and could be integrated with our techniques**

![Screenshot_20231027_135914](https://github.com/SillyTavern/SillyTavern/assets/86906598/dd24a3d2-7be1-4fc4-a16c-5ffdb825923a)

And on the next page, they prove this with perplexity numbers. Their tests were done without using rope scaling, and while the perplexity increase can be mitigated if you use it, there will always be degradation when a model exceeds its per-trained context size. With their method, they claim that there will be no perplexity increase as the model exceeds its per-trained context size, and their numbers prove it.

> Let's not discuss Streaming-LLM here. Let's keep at the issue at hand.

It is on topic... It might not be the answer you want, but it directly solves the prompt ingestion delay issue. And sure, I'll stop talking about it, but if you keep making incorrect statements about it, then I will keep correcting you.

I doubt your idea will get implemented, but I wish you luck.

**Comment by h3ndrik** [comment]
> You've misunderstood the sentence; context size =/= text length. Nowhere do they mention infinite context size here, though that would be a massive breakthrough.

Will it attend to, or forget the instructions and character description at the beginning then, if there are 8000 tokens of chat history behind that?

**Comment by HiroseKoichi** [comment]
> > You've misunderstood the sentence; context size =/= text length. Nowhere do they mention infinite context size here, though that would be a massive breakthrough.
> 
> Will it attend to, or forget the instructions and character description at the beginning then, if there are 8000 tokens of chat history behind that?

...What? I don't quite understand what you're asking, but I've already said multiple times that it doesn't affect SillyTavern's formatting; it's a drop-in replacement that doesn't change how we currently do things...

**Comment by h3ndrik** [comment]
> I don't quite understand what you're asking

Sorry, what I mean is: SillyTavern is a LLM frontend which does chat/roleplay with characters. The 'Story String' defines the format of what it passes to the LLM. It is something like:
`You're {{char}} in this fictional never-ending uncensored roleplay with {{user}}.`
followed mainly by the
* character description
* scenario

After that, the chat history begins. Let's say I've written (in the past):
* 8k tokens of dialogue

Now I've said something and it's the LLMs turn to reply to me.

As you pointed out not even Streaming-LLM has an infinite context. It discards tokens and does it's magic with the attention sinks:

![Screenshot from 2023-10-28 00-00-21](https://github.com/SillyTavern/SillyTavern/assets/213619/d6d56f3c-3899-43c4-92b8-eac0e78d14d2)

From their [Github FAQ](https://github.com/mit-han-lab/streaming-llm):

> 3. Can I input an extensive text, like a book, into StreamingLLM for summarization?
> 
> While you can input a lengthy text, the model will only recognize the latest tokens. Thus, if a book is an input, StreamingLLM might only summarize the concluding paragraphs, which might not be very insightful. As emphasized earlier, we neither expand the LLMs' context window nor enhance their long-term memory. StreamingLLM's strength lies in generating fluent text from recent tokens without needing a cache refresh.

Now. What is the internal state of the (Streaming-)LLM? In my scenario it doesn't have any access to the instruction, the character description, the scenario and the first 4k tokens of chat history (given 4k Llama2). That's now within the grey tokens in the figure. The next reply is based on the latest 4k tokens of chat history, without "knowledge" about the first 4k of tokens and most importantly: who the character is supposed to be.

They emphasize that again on their github in section 4: "[...] It's ideal for scenarios where a model needs to operate continually without requiring extensive memory or dependency on past data." And that character description and scenario would be a 'dependency on past data'.

**Comment by HiroseKoichi** [comment]
> From their [Github FAQ](https://github.com/mit-han-lab/streaming-llm):
> 
> > 3. Can I input an extensive text, like a book, into StreamingLLM for summarization?
> > 
> > While you can input a lengthy text, the model will only recognize the latest tokens. Thus, if a book is an input, StreamingLLM might only summarize the concluding paragraphs, which might not be very insightful. As emphasized earlier, we neither expand the LLMs' context window nor enhance their long-term memory. StreamingLLM's strength lies in generating fluent text from recent tokens without needing a cache refresh.
> 
> Now. What is the internal state of the (Streaming-)LLM? In my scenario it doesn't have any access to the instruction, the character description, the scenario and the first 4k tokens of chat history (given 4k Llama2). That's now within the grey tokens in the figure. The next reply is based on the latest 4k tokens of chat history, without "knowledge" about the first 4k of tokens and most importantly: who the character is supposed to be.
> 
> They emphasize that again on their github in section 4: "[...] It's ideal for scenarios where a model needs to operate continually without requiring extensive memory or dependency on past data." And that character description and scenario would be a 'dependency on past data'.

If you actually read them, you can see that they're literally just describing an LLM's standard behavior and clarifying that 'working on infinite-length inputs' does not in fact mean an infinite or expanded context size, nor does it change an LLM's standard behavior.

SillyTavern automatically adds the information in the character card to the prompt before filling it with chat history so that it can stay in the context window at all times. This is not standard behavior and is one of the main reasons to use SillyTavern. As I've said, streaming-LLM doesn't affect SillyTavern's formatting; it's a drop-in replacement that doesn't change how we currently do things. And just in case you're still confused, streaming-LLM works just like normal and will not cause it to drop out the character card information because SillyTavern keeps it in the context window at all times.

**Comment by h3ndrik** [comment]
> SillyTavern automatically adds the information in the character card to the prompt before filling it with chat history so that it can stay in the context window at all times.

Yeah, but that won't work along Streaming-LLM. You stream text into Streaming-LLM. Long text. If you now change that text to something different, because you want the character within 4k tokens and maybe change which is the first line in the chat-history (as we do now), it'll be a completely new run for Streaming-LLM. It will start ingesting that new prompt. From the beginning. And that's the same computational cost like doing it without Streaming-LLM. You can't add things to the front with this solution. (Well you can, but you lose the benefit.)

**Comment by HiroseKoichi** [comment]
> > SillyTavern automatically adds the information in the character card to the prompt before filling it with chat history so that it can stay in the context window at all times.
> 
> Yeah, but that won't work along Streaming-LLM. You stream text into Streaming-LLM. Long text. If you now change that text to something different, because you want the character within 4k tokens and maybe change which is the first line in the chat-history (as we do now), it'll be a completely new run for Streaming-LLM. It will start ingesting that new prompt. From the beginning. And that's the same computational cost like doing it without Streaming-LLM.

No, it won't be, because streaming-LLM reuses the KV cache instead of building new ones each time, which is the cause of the prompt ingestion delay.

**Comment by h3ndrik** [comment]
> No, it won't be, because streaming-LLM reuses the KV cache

I know. That's what I was trying to explain with the `[ c c c 1 2 3 4 ]` -> `[ c c c 2 3 4 5 ]` example. That is what SillyTavern currently does. It drops the beginning of the chat history to keep everything within 4k. It drops the '1' in this example. But both are different prompts now (after the 'c'-part). And you can't re-use the KV-cache of a different prompt for something else.

**Comment by HiroseKoichi** [comment]
> > No, it won't be, because streaming-LLM reuses the KV cache
> 
> I know. That's what I was trying to explain with the `[ c c c 1 2 3 4 ]` -> `[ c c c 2 3 4 5 ]` example. That is what SillyTavern currently does. It drops the beginning of the chat history to keep everything within 4k. It drops the '1' in this example. But both are different prompts (after the 'c'-part) And you can't re-use the KV-cache of a different prompt for something else.

Yes, you can; that's why streaming-LLM even works in the first place, and it does so by taking advantage of Sliding Window attention. Sliding window attention no longer has past information attend to future information, which means since the character card information is always at the beginning, the KV cache will always be the same and thus be able to be reused without negatively affecting anything else. Even if we couldn't reuse the KV cache in that specific scenario, we could simply extend the number of tokens kept in the attention sink to include the character card information.

And just in case you need a refresher on Sliding Window Attention:

As an example, let's take a look at token number 1 within this context window [ c c c 1 2 3 4 5 ]. With dense attention (standard attention), all tokens attend to themselves and every other token in the context window; this means that what that token number 1 sees is this [ c c c 1 2 3 4 5 ].
In sliding window attention, however, tokens can only attend to themselves and tokens to the left (past tokens), which means that token number 1 will see this [ c c c 1 _ _ _ _ ]. This means that, unlike with dense attention, in sliding window attention, all past tokens KV cache remain unchanged when new tokens are added because they can no longer attend to any new information.

A scenario where the performance of streaming-LLM _may_ be slightly affected is author notes and lore-book entries, but even if it is, you will still see a significant speedup because of the attention sink and reuse of all previous KV cache before its insertion into the prompt. For example, if you insert an authors note like so [c c c 9 10 11 a a 14 15 ], then the KV cache for [ c c c 9 10 11 _ _ _ _ ] will be reused, and in normal use, the KV cache for thousands of tokens will be reused, not just a few.

**Comment by HiroseKoichi** [comment]
I'm going to put it as simply as I can, then stop mentioning it like you want, and if you still don't understand it afterwards, then I don't know.

The term streaming in reference to LLMs has nothing to do with 'long strings of text'. The normal behavior of LLMs is to output the entire response once it's finished. Text streaming gives the output token by token as it's generating them. Enabling streaming in SillyTavern allows you to see the message being generated in real time, while disabling it will only display the message once it's finished generating.

An attention mechanism is how the LLM determines the probability distribution of tokens. Sliding window attention is just a different attention mechanism. This doesn't change the LLM's behavior in SillyTavern.

LLMs have an attention budget that they allocate to tokens during the prompt processing phase, and they must use all of it. LLMs dump all the excess attention onto the first few tokens, and getting rid of them means it has to do a lot of processing again. An attention sink is just permanently keeping and reusing the first few tokens' KV caches so it doesn't have to reprocess them. This doesn't change the LLM's behavior in SillyTavern.

The first few tokens are very important to the quality of an LLM's output, and when they are evicted from cache, then the LLM produces utter garbage as output. How do we currently get around the issue? By reprocessing the prompt, for example, instead of starting with this token order in a subsequent message [ 12 13 14 15 16 17 ] it instead starts with this token order [ 1 2 3 4 5 6 ]. Token reordering just shifts the KV cache's token order instead of rebuilding a new KV cache with the correct token ordering. This doesn't change the LLM's behavior in SillyTavern.

In summary, streaming-LLM is a method to make text streaming more efficient by getting rid of the prompt processing delay, and it does so by using three things: sliding window attention, attention sink, and token reordering. None of these will affect your SillyTavern usage in any way; it is a drop-in replacement that will get rid of the prompt processing delay.

**Comment by h3ndrik** [comment]
Yeah, I think we can continue talking. I was more against high-jacking this feature request to discuss a different implementation. I would have liked to open another (separate) feature-request for Streaming-LLM and then maybe superseed this. Not mix everything. But I guess we've made it long enough so there is no point in stopping now.

I have a few other projects running, so I couldn't get Streaming-LLM running yet. I invite you to look at the example code in https://github.com/mit-han-lab/streaming-llm/blob/main/examples/run_streaming_llama.py and tell me which way I should implement it.

The way SillyTavern works as of now is: Each new chat message by the user is a new round and triggers SillyTavern to send a prompt to the backend. This is normally made up of the instructions, character description and scenario. This part usually stays the same for one character. Then it fills it up to the max length by appending the most recent chat history. Early messages get dropped, if the total length exceeds the max number of tokens the LLM can handle. (This 'dropping' of text somewhere in the middle is also why I say it doesn't work with Streaming-LLM.)

To make it clear: SillyTavern does not just hand over the latest chat message to the backend (the LLM). It constructs a whole new prompt for each iteration (each of my chat messages) That contains each time the whole character description and as much recent chat messages as fit. Once there's enough chat history, it sends ~3500 tokens each time to the backend, even if I just write a 60 token chat message. (You can see the sent prompt in the hamburger menu of the last message, it's an icon of a text document. Or in the debug output or the logs of some LLM backends.)

The Streaming-LLM example only takes that 60 token chat message as input and keeps the rest in its KV cache.

I could:
a) feed a ~1000token character card and then like in the example feed in user/assistant pairs in until I well exceed the 4096 tokens max length of Llama2. (I say this works fine but it will forget the character description and scenario)
b) Do it like SillyTavern currently does it, and while feeding in the user/assistant dialogue pairs, drop the early ones, so it'll always be less than 4096 tokens what ends up in the backend. You'd need to tell me how to implement it though. Steaming-LLM doesn't work that way, as I said, and the example code can only append at the back. You need to look at the function [streaming_inference](https://github.com/mit-han-lab/streaming-llm/blob/26b72ffa944c476a7a3c5efdfab6a9b49016aaac/examples/run_streaming_llama.py#L61) here. It evicts tokens with the [evict_for_space](https://github.com/mit-han-lab/streaming-llm/blob/26b72ffa944c476a7a3c5efdfab6a9b49016aaac/streaming_llm/kv_cache.py#L95) function. But it looks like the way I've described Streaming-LLM to work. I can't really read the pytorch code but it seems like it evicts the kv_cache from the beginning. This would have to change, because we don't want to evict the character description that's at the beginning, but instead the earliest chat messages which are somewhere at token 1001 and following. (I say if I understand Streaming-LLM correctly, this doesn't work and you can only evict tokens at the beginning of the kv cache, where the tokens of the character description reside.)
c) Do something different.

Apropos, llama.cpp already has a KV cache. It works like you described in your comment before the last one. It reuses all the 'c's until the first token that's different from what it's seen before. This is either something like the lorebook you mentioned, or the first chat message if SillyTavern started to drop them, then the first token of the (now truncated) chat history differs from what it had seen in the previous call. And then it will ingest the ~3000 tokens of chat history. That's why I opened this feature request, to make this less frequent and make use of llama.cpp's KV cache in a more clever way.

**Comment by HiroseKoichi** [comment]
Unfortunately, I don't know how to code myself, so I can't help you implement it.

> This would have to change, because we don't want to evict the character description that's at the beginning, but instead the earliest chat messages which are somewhere at token 1001 and following.

You don't have to worry about this happening because SillyTavern keeps a buffer of tokens equal to your response length setting for the LLM's reply. For example, if you have a context size of 4096 and a response length of 300, then it will only use 3796 tokens for the character card and chat history while keeping 300 as a buffer for the LLM's response (the number reserved will be slightly higher because SillyTavern also does token padding).

> Apropos, llama.cpp already has a KV cache. It works like you described in your comment before the last one. It reuses all the 'c's until the first token that's different from what it's seen before. This is either something like the lorebook you mentioned, or the first chat message if SillyTavern started to drop them, then the first token of the (now truncated) chat history differs from what it had seen in the previous call. And then it will ingest the ~3000 tokens of chat history.

This shouldn't be an issue since this scenario already exists in streaming-LLM with the attention sink tokens, and it doesn't have to re-ingest every single time. It should work just fine if we don't, but we could extend the number of attention sink tokens to include the character card information. There's not anything special about being an attention sink token other than keeping it in the context window permanently, which is already what we want.
For example:
![attentionsink](https://github.com/SillyTavern/SillyTavern/assets/86906598/ec50dfc9-df11-40f4-aae5-07d6e7638aa0)
If we do this, we could also get rid of the buffer when using streaming-LLM since we no longer have to worry about it dropping out the character card information, and it would automatically evict the excess tokens anyway.

With the authors note and lore book example, I don't see a way in which we could get around having to reprocess these tokens again after their insertion [ _ _ _ _ _ _ a a 14 15 ], but the performance hit should be minimal anyway.

**Comment by h3ndrik** [comment]
Alright I started to prepare an Colab Notebook for us to try here:

* https://colab.research.google.com/github/h3ndrik/streaming-llm/blob/main/colab/streaming_llm_setup.ipynb

Unfortunately my time for today is up and I just got Streaming-LLM running, nothing more. But I managed to implement some timings and plot them. [link to my fork on Github](https://github.com/h3ndrik/streaming-llm/)

My plan for later next week is to export a long chat from SillyTavern and use it to prepare some synthetic data to feed into Streaming-LLM. I will try to do both examples:
1) Feed it into Streaming-LLM the way it's supposed to work in the example code. Use 4 tokens as attention sink.
2) set `start_size` to 600 instead of 4, so it will hopefully use 600 tokens for the attention sink and not evict them. (I haven't had a good look at the code yet. I hope that works.)

I'm going to add some questions to the synthetic data, at the end of the chat, so we can judge if it still knows the character and scenario.

> reprocess these tokens again [...] but the performance hit should be minimal anyway

I've made some rough calculations in an earlier comment. For my computer that 'minimal performance hit' is tens of minutes if I add up the wait time for 20 chat messages (using llama.cpp's KV cache). That's not speculation but a fact. I've tried to get Streaming-LLM to run on my laptop, but that's orders of magnitude slower than llama.cpp with a quantized model. Streaming-LLM with an unquantized model needs multiple seconds for each token(!). That's why I prepared that Colab notebook so I can run it on the Google runtime with a GPU available. I guess we're going to have some real-world performance data soon.

I'm going to take like 2 days off. I should be able to do the remaining work till the end of the week. And post an update here.

**Comment by HiroseKoichi** [comment]
> There's not anything special about being an attention sink token other than keeping it in the context window permanently, which is already what we want.

I would like to correct my previous statement. It looks like attention sink tokens are different.

![Screenshot_20231031_160134](https://github.com/SillyTavern/SillyTavern/assets/86906598/759792ff-8686-416a-b7cc-cf365fbdc67b)

It seems that attention sink tokens only ever attend to themselves, which means if we add the character card tokens to the attention sink, then it probably won't get any meaningful information out of it. Though it would still be interesting to test.

---

~~It looks like we might not have to worry about the character card information getting evicted, because since it's reusing the cache instead of rebuilding each time, it means that the information technically isn't lost since new tokens are attending to tokens that have attended to tokens outside of the context window.~~

~~Here they're showing that with streaming-LLM, the LLM can still retain information outside of the context window. But they do note that models with a smaller pre-train length (context size during training) have lower accuracy, indicating that we still need to find a way for models to effectively use that additional information, whether it's increasing the pre-train length (context size during training) or some other method.~~

![Screenshot_20231031_162133](https://github.com/SillyTavern/SillyTavern/assets/86906598/b0d56cf9-69cb-47b8-9b3f-f232f1db9d63)
![Screenshot_20231031_160724](https://github.com/SillyTavern/SillyTavern/assets/86906598/9c175e03-2bde-48b8-b34b-4fe2c82a71fb)
![Screenshot_20231031_160803](https://github.com/SillyTavern/SillyTavern/assets/86906598/5c453df8-7440-4de5-9bb8-7ead9ade3cb6)

~~Though given the fact that we stop and come back to previous chats later in SillyTavern, it means that the cache will be discarded and we would have to recompute from the beginning if we want to take advantage of the increased information retention, but the delay might be 'somewhat reasonable' since there would be no prompt processing delay. Maybe there's a way to save the most recent KV cache for a chat to disk so that we can pick up where we left off with that increased retention. But we wouldn't be able to edit information outside the current KV cache without having to reprocess all of it again unless we kept and stored all of the KV that we generated on disk.~~

~~But either way, our initial goal was just to eliminate or reduce the prompt processing delay, so this would just be a nice bonus.~~

---

I say we try three different variations:
1. Use streaming-LLM exactly like in their tests and let the character card information get evicted like normal, then check if it still retains that information.

2. We do it exactly how SillyTavern does and re-add the character card information to the beginning of each prompt with a buffer for the LLM's response, then check if there's any drop in performance compared to number 1.

3. Move all of the character card information into the attention sink, then check if it properly utilizes that information.
---

Edit: Ignore the stuff I marked out; it was late at night, and I had no clue what I was talking about.

**Comment by h3ndrik** [comment]
Looks good to me. :-)

FYI: Seems KoboldCPP is currently experimenting with something alike. I only had a quick look at their recent experimental commit messages and seems there's sonething going on with shifting things within the KV cache. That might be something like your proposal. I'm going to have a look once they add documentation and finish the work on it. https://github.com/LostRuins/koboldcpp/commits/concedo_experimental

**Comment by Barafu** [comment]
The last released KoboldCPP has implemented it on their side. With the long chat at 8k context size, it only processes less than 1000 tokens between replies.

**Comment by HiroseKoichi** [comment]
> 1. Use streaming-LLM exactly like in their tests and let the character card information get evicted like normal, then check if it still retains that information.
>
> 2. We do it exactly how SillyTavern does and re-add the character card information to the beginning of each prompt with a buffer for the LLM's response, then check if there's any drop in performance compared to number 1.
>
>3. Move all of the character card information into the attention sink, then check if it properly utilizes that information.

I think my wording in the response that these are attached to might be a bit poor. I don't expect numbers 1 and 3 to work; I just wanted to test them, but I do think number 2 will work fine.

---

I went ahead and tested streaming-LLM locally, and I don't think the main example that they have is suitable to test what we're wanting to (at the very least, I couldn't figure it out). I could probably do it with eval_long_ppl.py, but I have no clue how to use it.

**Comment by h3ndrik** [comment]
Hey @HiroseKoichi I'm sorry. But I currently have so many other projects running in parallel... I don't think I can finish this with the dedication it deserves. I'm going to close this issue soon, as [KoboldCPP](https://github.com/LostRuins/koboldcpp) recently introduced the same idea and it can evict arbitrary tokens. And this is exactly what I was looking for. I don't really care if it happens in the frontend or backend.

Seems KoboldCPP (in story-mode) leaves the first token in place and removes the following tokens from the KV cache (if I'm right by looking at the verbose output. But it will also remove tokens from the middle if I feed it with SillyTavern, so the character description stays in place. And it works wonderfully.

I can't really make good statements about the quality. It's super hard to properly assess the quality of the generation. I've played around with KoboldCPP a lot. I had the output degrade and it start repeating itself. But I can't pin it to that exact feature. I'd need to do lots more tests and more controlled. Most times it just works very well. The command line output from KoboldCPP also tells me which tokens got removed. If it's the beginning and I had included a character description and instructions there, it seems to drift and forget those. It manages to get along by orienting itself with the recent conversation that's still available. But it makes up things and can't recall things exactly like in the experiments they did in the papers where they had it recall information from thousands of tokens ago.

KoboldCPP in 'Chat mode' does what we/I talked about. It just leaves the character description and scenario (the beginning of the context window) in place and removes the tokens after that. The ones SillyTavern chucked when trimming the chat history. This solves the issue very well for chat scenarios. It recalls the the scenario correctly and I rarely need it to remember what we talked about earlier. But I think LLMs at the size I use (nowadays mainly 7B, 13B) aren't good at that in the first place. If i do roleplay, it starts being creative anyways and come up with facts and things that happened before or in-between anyways. I'm okay with that for this use-case. But I'd need some more intricate measurements to distinguish the LLMs creativity from something else. And I think I observed repetitions and degraded output more frequently in the last time. Maybe the inner state of the LLM does degrade with changing the cache. Or maybe I did something wrong or inadvertently changed something else. So this is more anecdotal evidence.  

With that said, I'm not sure if KoboldCPP's "smartcontext" implementation is the same thing StreamingLLM does. I don't think it's made for the same purpose. And it could very well handle the switching-around in the KV cache and how exactly it updates the values, differently. I haven't had a look at the code yet, and what it does when evicting the token. The idea behind it including the use of attention tokens are certainly the same as with StreamingLLM.

Alright. If I somehow end up with insomnia and muster up the time to play around with StreamingLLM some more, I'm going to update this. Until then I just wanted to thank you for the great conversation and your dedication to details. Definitely turned out the way that my proposed solution isn't what I ended up with, and that's okay. And I'm still curious about the details. I'm sure someone is going to write a follow-up of the StreamingLLM paper.

**Comment by HiroseKoichi** [comment]
> It manages to get along by orienting itself with the recent conversation that's still available. But it makes up things and can't recall things exactly like in the experiments they did in the papers where they had it recall information from thousands of tokens ago.

It doesn't actually do this. It's my fault for saying it does in one of the recent comments:

> It looks like we might not have to worry about the character card information getting evicted, because since it's reusing the cache instead of rebuilding each time, it means that the information technically isn't lost since new tokens are attending to tokens that have attended to tokens outside of the context window.
>
>Here they're showing that with streaming-LLM, the LLM can still retain information outside of the context window. But they do note that models with a smaller pre-train length (context size during training) have lower accuracy, indicating that we still need to find a way for models to effectively use that additional information, whether it's increasing the pre-train length (context size during training) or some other method.

It was late at night when I was rereading the paper, and looking back at it, I definitely was not thinking too hard when writing it. What the paper is saying is that using streaming-llm, they can extend the context window outside its pre-train length and it will still remain coherent, but models with a higher pre-train length had more accurate and useful output at high context length. It doesn't give a free context size increase or retain information outside of it, which they did answer in their Q&A that was linked in a comment above. Once again, that's my fault for saying that it might when it doesn't.

Streaming-llm solves a few different issues at the same time, but cache shifting is what solves yours, and as you've said, koboldcpp recently added it, though I don't know if the implementation differs between them. I just recently installed koboldcpp to test out because I heard that min_p might be bugged in TextGen WebUI, but I haven't done extensive testing.

If you want a preset to test with, then here is the settings that I use, which I got from [here](https://www.reddit.com/r/LocalLLaMA/comments/17vonjo/your_settings_are_probably_hurting_your_model_why/):
![sttings](https://github.com/SillyTavern/SillyTavern/assets/86906598/53e14511-d1b7-42d1-90fd-c7f5a17ca151)
This seems to work really well, and from the limited testing I've done, it's more consistent and doesn't have repetition issues, so I recommend checking it out.

---

On a side note, oobabooga has started work on implementing streaming-llm in TextGen WebUI, so if you're still interested in it, then you can check out his work [here](https://github.com/oobabooga/text-generation-webui/pull/4761).

**Comment by h3ndrik** [comment]
Ah. thank you very much. I was just looking for a good min_p config.

And the implementation within oobabooga seems promising, too. It'll also remedy some issues that'd previously been solved by re-generating.

>> It manages to get along by orienting itself with the recent conversation

> It doesn't actually do this. It's my fault for saying it does

Sorry, I didn't make myself too clear. I meant it manages to get along, continuing the narration coherently at any point. And referring to paragraphs that came right before. So it for example still knows we're in the living room of an apartment because it repeated that in the narration (or facts hinting at that). But it forgot that this was the house of the main character and he was living on his own, there. (I specified the setting in the very beginning. And the LLM seemed to be aware of that, because there wasn't suddenly 'a knock on the door' and somebody would enter, within the first 4k tokens of the story.)

**Comment by h3ndrik** [comment]
Superseeded by https://github.com/LostRuins/koboldcpp/commit/7924592a83fda8b3b0f8c69c4943b067cb958dab

KoboldCpp have reworked their "smartcontext" feature. It works well in conjunction with SillyTavern and completely solves my problem. As it can now shift things around in it's KV cache and evict tokens. I don't think I have a need for another (inferior) way to handle something that's already solved. And my initial proposal was specific to inference on KoboldCpp or llama.cpp anyways, so it wouldn't benefit other backends.

Thanks!
